stlab/concurrency/default_executor.hpp

```diff
@@ -178,135 +221,188 @@ class task_system {
     }
 };
 
+/**************************************************************************************************/
+/**************************************************************************************************/
+
 #elif STLAB_TASK_SYSTEM == STLAB_TASK_SYSTEM_PORTABLE
 
-/**************************************************************************************************/
 
 class notification_queue {
     using lock_t = std::unique_lock<std::mutex>;
     std::deque<task<void()>> _q;
-    bool _done{false};
-    std::mutex _mutex;
-    std::condition_variable _ready;
+    std::mutex _queue_mutex;
+    std::condition_variable* _ready{nullptr};
 
 public:
+    void set_context(std::condition_variable& cv) {
+        _ready = &cv;
+    }
+
     bool try_pop(task<void()>& x) {
-        lock_t lock{_mutex, std::try_to_lock};
+        lock_t lock{_queue_mutex, std::try_to_lock};
         if (!lock || _q.empty()) return false;
         x = std::move(_q.front());
         _q.pop_front();
         return true;
     }
 
+    bool pop(task<void()>& x) {
+        lock_t lock{_queue_mutex};
+        if (_q.empty()) return false;
+        x = std::move(_q.front());
+        _q.pop_front();
+        return true;
+    }
+
+
     template <typename F>
     bool try_push(F&& f) {
         {
-            lock_t lock{_mutex, std::try_to_lock};
+            lock_t lock{_queue_mutex, std::try_to_lock};
             if (!lock) return false;
             _q.emplace_back(std::forward<F>(f));
         }
-        _ready.notify_one();
-        return true;
-    }
-
-    void done() {
-        {
-            lock_t lock{_mutex};
-            _done = true;
-        }
-        _ready.notify_all();
-    }
-
-    bool pop(task<void()>& x) {
-        lock_t lock{_mutex};
-        while (_q.empty() && !_done)
-            _ready.wait(lock);
-        if (_q.empty()) return false;
-        x = std::move(_q.front());
-        _q.pop_front();
+        _ready->notify_all();
         return true;
     }
 
     template <typename F>
     void push(F&& f) {
         {
-            lock_t lock{_mutex};
+            lock_t lock{_queue_mutex};
             _q.emplace_back(std::forward<F>(f));
         }
-        _ready.notify_one();
+        _ready->notify_all();
     }
 };
 
 /**************************************************************************************************/
 
-class task_system {
+class priority_task_system {
+    using lock_t = std::unique_lock<std::mutex>;
+
     const unsigned _count{std::thread::hardware_concurrency()};
-    std::vector<std::thread> _threads;
-    std::vector<notification_queue> _q{_count};
+
+    struct thread_context
+    {
+        std::mutex mutex;
```

Comment:
I believe stlab convention is to prefix member variables with `_`.

Reply:
I will fix this. I am used to our project's style, where pure value types don't have '_'.

Reply:
Done